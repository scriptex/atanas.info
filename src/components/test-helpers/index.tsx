import * as React from 'react';
import { act } from 'react-dom/test-utils';
import { render, RenderResult, waitFor } from '@testing-library/react';

import * as shared from '../../scripts/shared';

export const mockUseEffect = (): void => {
	let useEffect: jest.Mock<React.EffectCallback>;

	const mockedUseEffect = () => {
		useEffect.mockImplementationOnce((f: () => React.EffectCallback) => f());
	};

	beforeEach(() => {
		useEffect = jest.spyOn(React, 'useEffect') as unknown as jest.Mock<React.EffectCallback>;

		mockedUseEffect();
	});
};

export const snapshotTest = (
	Component: React.FC<Readonly<React.ComponentProps<any>>>,
	buttonSelector?: string
): void => {
	const { name } = Component;

	jest.spyOn(shared, 'formatDate');

	(shared.formatDate as jest.Mock).mockImplementation(() => 'Mock date');

	mockUseEffect();

	afterAll(() => {
		jest.clearAllMocks();
	});

	describe(`${name} component`, () => {
		it(`Should render the ${name} component`, async () => {
			let result!: RenderResult;

			await act(async () => {
				result = await waitFor(() => render(<Component />));
			});

			const { asFragment, container } = result;

			expect(asFragment()).toMatchSnapshot();

			if (buttonSelector) {
				act(() => {
					container.querySelectorAll<HTMLElement>(buttonSelector).forEach(button => {
						if (!button.hasAttribute('download')) {
							button.click();

							expect(asFragment()).toMatchSnapshot();
						}
					});
				});
			}
		});

		it(`Should unmount the ${name} component`, async () => {
			let result!: RenderResult;

			await act(async () => {
				result = await waitFor(() => render(<Component />));
			});

			const { unmount, asFragment } = result;

			unmount();

			expect(asFragment()).toMatchSnapshot();
		});
	});
};
