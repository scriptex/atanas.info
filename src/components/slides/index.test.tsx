import { Slides } from '.';
import { snapshotTest } from '../test-helpers';

snapshotTest(Slides);
