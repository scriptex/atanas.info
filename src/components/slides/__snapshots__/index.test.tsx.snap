// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slides component Should render the Slides component 1`] = `
<DocumentFragment>
  <main
    class="o-main "
  >
    <section
      class="c-section c-section--slides fullsize-background"
      id="slides"
      style="background-image: url(images/temp/presentation.jpg);"
    >
      <div
        class="o-shell"
      >
        <header
          class="c-section__header"
        >
          <h2>
            Slides
          </h2>
        </header>
        <nav
          class="c-section__nav"
        >
          <ul>
            <li
              class="current"
            >
              <button
                aria-label="Switch to EcmaScript Features"
                class="c-btn "
                type="button"
              >
                EcmaScript Features
              </button>
            </li>
            <li
              class=""
            >
              <button
                aria-label="Switch to Do more with Sublime Text"
                class="c-btn "
                type="button"
              >
                Do more with Sublime Text
              </button>
            </li>
          </ul>
        </nav>
        <div
          class="c-section__body"
        >
          <div
            class="c-section__frame current"
          >
            <div
              class="c-loading"
            >
              <div
                class="c-loader"
              />
            </div>
            <iframe
              allowfullscreen=""
              loading="lazy"
              src="https://docs.google.com/presentation/d/e/2PACX-1vSA8qN0_3lb0YWDYJheJ3q0AAsXWlLLpP86pnmV0KWfPo_2G-SSdSK--GVRyJbwT8_kWyJ0TZxFoaH8/embed?start=false&loop=false&delayms=3000"
              title="Presentation about the EcmaScript language history and features"
            />
          </div>
          <div
            class="c-section__frame"
          >
            <div
              class="c-loading"
            >
              <div
                class="c-loader"
              />
            </div>
            <iframe
              allowfullscreen=""
              loading="lazy"
              src="https://docs.google.com/presentation/d/e/2PACX-1vTivcVmPyjYPdWgoJMChe_rrVyZEIw4ZgF37lCOlxUUm7DchxhMJouEV_BcrH6LDL8kiNmRSIWkiSxK/embed?start=false&loop=false&delayms=3000"
              title="Presentation about improving workflow using Sublime Text editor"
            />
          </div>
        </div>
        <div
          class="c-section__actions"
        >
          <button
            class="c-btn "
            type="button"
          >
            Get in touch
          </button>
          <ul
            class="c-section__actions-list"
          >
            <li>
              <button
                class="c-btn c-section__actions-close"
                type="button"
              >
                Close
              </button>
            </li>
            <li>
              <button
                class="c-btn "
                type="button"
              >
                <svg
                  class="c-section__actions-icon"
                >
                  <use
                    xlink:href="#svg-email"
                  />
                </svg>
                <span>
                  Contact me
                </span>
                <strong>
                  directly
                </strong>
              </button>
            </li>
            <li>
              <a
                class="c-btn"
                href="https://app.usebraintrust.com/talent/782/"
                rel="noopener noreferrer"
                target="_blank"
              >
                <svg
                  class="c-section__actions-icon"
                >
                  <use
                    xlink:href="#svg-braintrust"
                  />
                </svg>
                <span>
                  Hire me on
                </span>
                <strong>
                  Braintrust
                </strong>
              </a>
            </li>
            <li>
              <a
                class="c-btn"
                href="https://www.toptal.com/resume/atanas-atanasov"
                rel="noopener noreferrer"
                target="_blank"
              >
                <svg
                  class="c-section__actions-icon"
                >
                  <use
                    xlink:href="#svg-toptal"
                  />
                </svg>
                <span>
                  Hire me on
                </span>
                <strong>
                  Toptal
                </strong>
              </a>
            </li>
          </ul>
        </div>
      </div>
    </section>
  </main>
</DocumentFragment>
`;

exports[`Slides component Should render the Slides component 2`] = `
<DocumentFragment>
  <main
    class="o-main "
  >
    <section
      class="c-section c-section--slides fullsize-background"
      id="slides"
      style="background-image: url(images/temp/presentation.jpg);"
    >
      <div
        class="o-shell"
      >
        <header
          class="c-section__header"
        >
          <h2>
            Slides
          </h2>
        </header>
        <nav
          class="c-section__nav"
        >
          <ul>
            <li
              class="current"
            >
              <button
                aria-label="Switch to EcmaScript Features"
                class="c-btn "
                type="button"
              >
                EcmaScript Features
              </button>
            </li>
            <li
              class=""
            >
              <button
                aria-label="Switch to Do more with Sublime Text"
                class="c-btn "
                type="button"
              >
                Do more with Sublime Text
              </button>
            </li>
          </ul>
        </nav>
        <div
          class="c-section__body"
        >
          <div
            class="c-section__frame current"
          />
          <div
            class="c-section__frame"
          />
        </div>
        <div
          class="c-section__actions"
        >
          <button
            class="c-btn "
            type="button"
          >
            Get in touch
          </button>
          <ul
            class="c-section__actions-list"
          >
            <li>
              <button
                class="c-btn c-section__actions-close"
                type="button"
              >
                Close
              </button>
            </li>
            <li>
              <button
                class="c-btn "
                type="button"
              >
                <svg
                  class="c-section__actions-icon"
                />
                <span>
                  Contact me
                </span>
                <strong>
                  directly
                </strong>
              </button>
            </li>
            <li>
              <a
                class="c-btn"
                href="https://app.usebraintrust.com/talent/782/"
                rel="noopener noreferrer"
                target="_blank"
              >
                <svg
                  class="c-section__actions-icon"
                />
                <span>
                  Hire me on
                </span>
                <strong>
                  Braintrust
                </strong>
              </a>
            </li>
            <li>
              <a
                class="c-btn"
                href="https://www.toptal.com/resume/atanas-atanasov"
                rel="noopener noreferrer"
                target="_blank"
              >
                <svg
                  class="c-section__actions-icon"
                />
                <span>
                  Hire me on
                </span>
                <strong>
                  Toptal
                </strong>
              </a>
            </li>
          </ul>
        </div>
      </div>
    </section>
  </main>
</DocumentFragment>
`;
